### PR TITLE
Multivariate Detrending

### LINKED ISSUE #1030: Make transformer for detrending support multivariate series
**Is your feature request related to a problem? Please describe.**
For the EEG applications we need a transformer that detrends each channel. Currently the detrending transformer only supports univariate time series.

**Describe the solution you'd like**
Detrend each univariate series separately

### PR DESCRIPTION
**Pull request by SveaMeyer13**
#### Reference Issues/PRs

Fixes #1030 . See also #1038 .

#### What does this implement/fix? Explain your changes.

This would be a quick fix to #1030 as it adds functionality for multivariate series to the detrender. As discussed with @mloning I took inspiration from how this was done for the imputer. But I also like the idea to have compositions for multivariate series as it is described in #1038

### CODE REVIEW

**Comment by fkiraly** [comment]
Really nice!

What I would suggest though to address: the columns of all the data frames passed need not be in the same order.
So I would suggest:

* check whether all columns appearing later than `fit` have already been seen
* index `forecaster_` copies by column name, using a `dict`, not by integer index, that makes lookup easier

**Review by fkiraly — CHANGES_REQUESTED** [review]
see above - suggest to robustify this with regards to:

* columns in different order in different calls after `fit`
* different set of columns passed

**Comment by SveaMeyer13** [comment]
Thank you for your feedback @fkiraly - it's really helpful. 

Just to make sure I understood correctly: 

a) currently, forecaster_ is a list of forecasters and I will change it to a dict that contains the column name as keys and the forecasters as values

b) I will check if a column has already been seen, by checking whether the forecaster that belongs to that column is fitted, right?

**Comment by fkiraly** [comment]
>a) currently, forecaster_ is a list of forecasters and I will change it to a dict that contains the column name as keys and the forecasters as values

That's precisely what I would suggest, yes.

>b) I will check if a column has already been seen, by checking whether the forecaster that belongs to that column is fitted, right?

What I would have done is simply check whether the list of columns is a sub-set of the `dict`'s keys, which should work if those keys are generated in `fit`. Alternatively, you could store the columns you see in `fit` in an attribute and use that.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks good, almost ready to go.

The one thing I would recommend to change is the error message `"not all columns are known"` - this is vague and not very helpful. Not all columns *of what*, and known *in which respect*?

Suggest: `Z` contains columns that have not been seen in `fit` (maybe even print those columns by including them in the error msg), and change the error to `ValueError` (since it's about the contents/values of `Z` - it is locally about a key, but this should be in reference to the inputs, not local objects).

**Comment by SveaMeyer13** [comment]
@fkiraly, thanks for the feedback! I'll take care of it

File: sktime/transformations/series/detrend/_detrend.py [hunk 1]
```diff
@@ -107,10 +117,22 @@ def transform(self, Z, X=None):
...
-        z = check_series(Z, enforce_univariate=True)
+        z = check_series(Z)
         fh = ForecastingHorizon(z.index, is_relative=False)
-        z_pred = self.forecaster_.predict(fh, X)
-        return z - z_pred
+
+        # multivariate
+        if isinstance(z, pd.DataFrame):
+            # check if all columns are known
+            if not all(k in self.forecaster_ for k in z.columns):
```
Comment (fkiraly) [inline sktime/transformations/series/detrend/_detrend.py:126]: for the check or printing missing columns, you could do sth like `set(z.columns).difference(set(self.forecaster_.keys()))`; for readability you could first define variables `Z_fit_keys = set(self.forecaster_.keys())`, `Z_new_keys = ...`

Comment (SveaMeyer13) [reply sktime/transformations/series/detrend/_detrend.py]: Thanks! I implemented it as you suggested

**Review by fkiraly — APPROVED** [review]
looks great!

We can merge once the tests complete.
Do you want to have a look too, @mloning?

**Comment by mloning** [comment]
You can merge if you think it's all good, happy to take a look if you're unsure.

**Comment by aiwalter** [comment]
@SveaMeyer13 I just tested this and I got a warning:
```
C:\...\sktime\sktime\transformations\series\detrend\_detrend.py:136: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead
See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  z[colname] = z[colname] - z_pred
```
Could you have a look into it?
___
Martin Walter <martin_friedrich.walter@daimler.com>, Mercedes-Benz AG on behalf of Daimler TSS GmbH.
https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md

**Comment by SveaMeyer13** [comment]
Thanks, @aiwalter! 

For some reason I can not reproduce this warning when running the following code:

```python
y, X = base.load_longley()
transformer = Detrender(forecaster=PolynomialTrendForecaster(degree=1))
y_hat = transformer.fit_transform(X)
print(y_hat)
```

But I do see the problem and I'm looking into it. 

For now, I think I should have made a deep copy of z in the transform/inverse-transform function before applying changes to it.

**Comment by aiwalter** [comment]
Will post the code tomorrow to reproduce

**Comment by aiwalter** [comment]
@SveaMeyer13 seems the warning was caused by sth in my `ForecastingPipeline` which I have fixed now. So all is fine with your PR, was my mistake! I just used the new `Detrender` there for some tests...
___
Martin Walter <martin_friedrich.walter@daimler.com>, Mercedes-Benz AG on behalf of Daimler TSS GmbH.
https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md

**Pull request merged**